File: src/plugins/utils/typescript/types-info.ts
import * as fs from "fs";
import path from "path";

// import { parse } from "@typescript-eslint/typescript-estree";
// import createCsvWriter from "csv-writer";

/**
 * This script sets up the basic structure but leaves the implementation of analyzeTypeScriptFile function open. This function should analyze the AST of each TypeScript file to extract type information. This is a complex task because we need to understand the structure of the TypeScript AST and identify how different types (interfaces, types, classes, etc.) are represented in it.
 */

interface TypeInfo {
  fileName: string;
  typeName: string;
  location: string;
  // Other details as needed
}

const projectPath = "/path/to/the/project";
const csvFilePath = "typescript-types.csv";

// const csvWriter = createCsvWriter.createArrayCsvWriter({
//   header: ["File Name", "Type Name", "Location"],
//   path: csvFilePath,
// });

function findTSFiles(dir: string, files_: string[] = []) {
  // biome-ignore lint/complexity/noForEach: <explanation>
  fs.readdirSync(dir).forEach((file) => {
    const filePath = path.join(dir, file);
    if (fs.statSync(filePath).isDirectory()) {
      findTSFiles(filePath, files_);
    } else if (filePath.endsWith(".ts") || filePath.endsWith(".tsx")) {
      files_.push(filePath);
    }
  });
  return files_;
}

function analyzeTypeScriptFile(file: string): TypeInfo[] {
  const content = fs.readFileSync(file, "utf8");
  // const ast = parse(content, {
  //   loc: true,
  //   // Other relevant options as needed
  // });

  // TODO: Analyze AST and extract type information
  // This part requires in-depth knowledge of TypeScript AST
  return []; // Return array of TypeInfo
}

function main() {
  const tsFiles = findTSFiles(projectPath);
  // eslint-disable-next-line sonarjs/no-unused-collection
  const allTypeInfo: TypeInfo[] = [];

  // biome-ignore lint/complexity/noForEach: <explanation>
  tsFiles.forEach((file) => {
    const typesInFile = analyzeTypeScriptFile(file);
    allTypeInfo.push(...typesInFile);
  });

  // csvWriter
  //   .writeRecords(
  //     allTypeInfo.map((typeInfo) => [
  //       typeInfo.fileName,
  //       typeInfo.typeName,
  //       typeInfo.location,
  //     ]),
  //   )
  //   .then(() => console.log("CSV file was written successfully"));
}

main();

File: src/plugins/utils/typescript/types-info.py
import os
import csv
from typing import List, Dict, Any

# This script provides a framework, but the crucial part, analyze_typescript_file, where we parse and extract the type information from TypeScript files, is not implemented. This requires a more sophisticated approach, possibly using a TypeScript parser or AST (Abstract Syntax Tree) analyzer in Python. The possible library for this is typescript-estree, which we can use via a Node.js script and then integrate its output into our Python script.
# Also, the information we might want to extract about each type could include:
# Type name | File where it's defined | Line number | Properties of the type (if it's an object)
# Extends or implements other types (inheritance/interfaces) | Usage count across the project

def find_ts_files(directory: str) -> List[str]:
    ts_files = []
    for root, dirs, files in os.walk(directory):
        for file in files:
            if file.endswith('.ts') or file.endswith('.tsx'):
                ts_files.append(os.path.join(root, file))
    return ts_files

def analyze_typescript_file(file_path: str) -> List[Dict[str, Any]]:
    # TODO: Implement the logic to analyze the typescript file
    # and extract type information
    return []

def write_to_csv(data: List[Dict[str, Any]], output_file: str):
    with open(output_file, mode='w', newline='', encoding='utf-8') as file:
        writer = csv.DictWriter(file, fieldnames=data[0].keys())
        writer.writeheader()
        writer.writerows(data)

def main():
    project_directory = '/path/to/your/nextjs/project'
    output_csv = 'typescript_analysis.csv'

    all_types = []
    for ts_file in find_ts_files(project_directory):
        types_in_file = analyze_typescript_file(ts_file)
        all_types.extend(types_in_file)

    write_to_csv(all_types, output_csv)

if __name__ == '__main__':
    main()


